src/lib/optimizer/strategy/join_predicate_ordering_rule.cpp

```diff
@@ -0,0 +1,69 @@
+#include "expression/abstract_predicate_expression.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "statistics/cardinality_estimator.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+void JoinPredicateOrderingRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  // Check if this is a multi predicate join.
+  if (node->type != LQPNodeType::Join || node->node_expressions.size() <= 1) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  const auto join_mode = std::static_pointer_cast<JoinNode>(node)->join_mode;
+
+  DebugAssert(cost_estimator, "JoinOrderingRule requires cost estimator to be set");
+  const auto caching_cardinality_estimator = cost_estimator->cardinality_estimator->new_instance();
+
+  // Estimate join selectivity of a predicate by getting cardinalities for a join node joining only on that one predicate.
+  //  “The join selectivity is the ratio “number of tuples in the result/number of tuples in the cartesian product.”
+  //  See http://www.vldb.org/journal/VLDBJ6/70060191.pdf for more infos on join selectivity.
+  auto predicate_cardinalities = std::unordered_map<std::shared_ptr<AbstractExpression>, Cardinality>{};
+  for (const auto& predicate : node->node_expressions) {
+    const auto single_predicate_join = JoinNode::make(join_mode, predicate);
+    single_predicate_join->set_left_input(node->left_input());
+    single_predicate_join->set_right_input(node->right_input());
+
+    predicate_cardinalities[predicate] = caching_cardinality_estimator->estimate_cardinality(single_predicate_join);
+  }
+
+  // Sort predicates by ascending join selectivity.
+  std::sort(node->node_expressions.begin(), node->node_expressions.end(),
+            [&](const std::shared_ptr<AbstractExpression>& a, const std::shared_ptr<AbstractExpression>& b) {
+              return predicate_cardinalities[a] < predicate_cardinalities[b];
+            });
+
+  // Semi and anti joins are currently only implemented by hash joins. These need an equals comparison as the primary
+  // join predicate. Check that one exists and move it to the front.
+  if (join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue || join_mode == JoinMode::AntiNullAsFalse) {
+    auto first_equals_predicate =
+        std::find_if(node->node_expressions.begin(), node->node_expressions.end(),
+                     [](const std::shared_ptr<AbstractExpression>& expression) {
+                       return std::static_pointer_cast<AbstractPredicateExpression>(expression)->predicate_condition ==
+                              PredicateCondition::Equals;
+                     });
+
+    Assert(first_equals_predicate != node->node_expressions.end(),
+           "Semi/anti joins require at least one equals predicate at the moment.");
+
+    while (first_equals_predicate != node->node_expressions.begin()) {
```

Comment:
Doc this loop. Is there really no more STL-like way of bubbling up an element through a vector? This looks ugly.

Reply:
Agreed, I will have another look.

Reply:
std::rotate does the trick.